### PR TITLE
Drive train gyro

### PR DESCRIPTION
**Pull request by chloe6248**

### CODE REVIEW

File: src/org/usfirst/frc/team2399/robot/RobotMap.java [hunk 1]
```diff
@@ -63,9 +64,14 @@
...
-	public static final double DISTANCE_INCREMENT = 0;
-	public static final double ANGLE_P = 0;
+	public static final double DRIVE_DISTANCE_DECREMENT = 0;
+	public static final double DRIVE_DISTANCE_INCREMENT = 0;
+	public static final double DRIVE_ANGLE_DECREMENT = 0;
+	public static final double DRIVE_ANGLE_INCREMENT = 0;
+	public static double DRIVE_ANGLE_P = 0;
+	public static final double DRIVE_ANGLE_ERROR = 0;
+	public static final double DRIVE_MIXED_LINEAR = 0;
+	public static final double DRIVE_MIXED_ANGULAR = 0;
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/RobotMap.java]: I think 73-74 should have some description of what they do. (edit: I wrote this before you explained what 'mixed' meant, but I think I stand by it. 'mixed' is a good, short word for what it intends to describe, but a little note somewhere would be helpful)

File: src/org/usfirst/frc/team2399/robot/commands/DriveAnglePAdjustment.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package org.usfirst.frc.team2399.robot.commands;
+
+import org.usfirst.frc.team2399.robot.Robot;
+import org.usfirst.frc.team2399.robot.subsystems.DriveTrain;
+
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/commands/DriveAnglePAdjustment.java]: A brief Javadoc should be filled out before merge. I'm not totally sure what these classes are for; I'm guessing the PID-loop tuning?

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java [hunk 3]
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
 	 * the P to the current P + the increment/decrement constant
 	 */
 	
-	public void incrementDistanceConstant(CANTalon currentTalon){
-		double currentPConstant = currentTalon.getP();
-		currentTalon.setP(currentPConstant += RobotMap.DISTANCE_INCREMENT);
+	public void incrementRightDistancePConstant(){
+		double currentPConstant = rightFrontTalon.getP();
+		rightFrontTalon.setP(currentPConstant += RobotMap.DRIVE_DISTANCE_INCREMENT);
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java:240]: Separation of concerns = great! These are in the right place. May I suggest making this one line, though? You create the `currentPConstant` variable, then use it only once; you could simply call the function inline. Same goes for the next few methods as well.

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java [hunk 4]
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
...
+		return rightFrontTalon.getP();
+	}
+	
+	public double returnLeftDistanceConstant(){
+		return leftFrontTalon.getP();
+	}
+	
+	public double getCurrentAngle()
+	{
+		return Navx.getYaw();
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: I'd like a javadoc that describes the return value a bit more. E.g. if the method returns `90.0`, what does that actually mean? Is it relative to the initial position? Is it clockwise or CCW? (If this is actually an obvious thing that I'm just missing, let me know and I'll clear this comment)

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java [hunk 5]
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
...
+	public double getDesiredAngle()
+	{
+		return desiredAngle;
+	}
+	
+	public double calculateAngleError()
+	{
+		double newDesiredAngle;
+		
+		if(getCurrentAngle()-180 >= getDesiredAngle())
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Note: After chatting with team, I'd like a group activity during which we simplify this method. It does work, but I am like 90% certain it can be done more simply.

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java [hunk 6]
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
...
+		{
+			newDesiredAngle = getDesiredAngle();
+		}
+		
+		return newDesiredAngle - getCurrentAngle();
+	}
+	
+	public void moveToAngle()
+	{
+		double pOutput = calculateAngleError() * anglePConstant;
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Hmm, I don't think this is how p-constants are supposed to work. Perhaps I am misunderstanding, or this is a different P from the PID loop, but that P should be fed into the Talons.

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java [hunk 7]
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
...
+		driveRightVelocity(-pOutput);
+		driveLeftVelocity(pOutput);
+	}
+	
+	public boolean isDriveAngleFinished()
+	{
+		if(Math.abs(calculateAngleError()) <= RobotMap.DRIVE_ANGLE_ERROR)
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: You can shorten this to just say `return Math.abs(calculateAngleError()) <= RobotMap.DRIVE_ANGLE_ERROR`

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java [hunk 8]
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
...
+	public double getAnglePConstant()
+	{
+		return anglePConstant;
 	}
 	
-	public double returnDistanceConstant(CANTalon currentTalon){
-		return currentTalon.getP();
+	public void driveAtAngleToDistance()
+	{
+		double anglePOutput = calculateAngleError() * anglePConstant;
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Same question as above r/e use of p-constant

File: .classpath [hunk 9]
```diff
@@ -7,5 +7,6 @@
 	<classpathentry kind="var" path="cscore" sourcepath="cscore.sources"/>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
 	<classpathentry kind="var" path="USERLIBS_DIR/TalonSRXLibJava.jar"/>
+	<classpathentry kind="var" path="navx-mxp"/>
```
Comment (jameshochadel) [inline .classpath]: Just brainstorming, but it might not be a bad idea to have a "setting up the 2017-Season project" document somewhere; a README in the project repo would be perfect for this! (It'd be pretty simple; just point them to the WPI 2017 System Update stuff and the notes you made in the journals for installing this)

File: src/org/usfirst/frc/team2399/robot/commands/DriveAnglePAdjustment.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+ */
+public class DriveAnglePAdjustment extends Command {
+	
+	private DriveTrain driveTrain = Robot.driveTrain;
+	private boolean isIncrementing;
+	
+
+    public DriveAnglePAdjustment(boolean isIncrementing) {
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/commands/DriveAnglePAdjustment.java]: Good param naming!

File: src/org/usfirst/frc/team2399/robot/commands/DriveDistancePAdjustment.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
+package org.usfirst.frc.team2399.robot.commands;
+
+import org.usfirst.frc.team2399.robot.Robot;
+import org.usfirst.frc.team2399.robot.subsystems.DriveTrain;
+
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/commands/DriveDistancePAdjustment.java]: Same as last one - brief Javadoc plz

File: src/org/usfirst/frc/team2399/robot/commands/DriveDistancePAdjustment.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    	this.isIncrementing = isIncrementing;
+    }
+
+    // Called just before this Command runs the first time
+    protected void initialize() {
+    }
+
+    // Called repeatedly when this Command is scheduled to run
+    protected void execute() {
+    	if(isIncrementing == true)
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/commands/DriveDistancePAdjustment.java]: Line 29 can be simpler! (I will leave the simplification as an exercise to the reader)

File: src/org/usfirst/frc/team2399/robot/commands/DriveAnglePAdjustment.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    	this.isIncrementing = isIncrementing;
+    }
+
+    // Called just before this Command runs the first time
+    protected void initialize() {
+    }
+
+    // Called repeatedly when this Command is scheduled to run
+    protected void execute() {
+    	if(isIncrementing == true)
```
Comment (jameshochadel) [inline src/org/usfirst/frc/team2399/robot/commands/DriveAnglePAdjustment.java]: Line 29 can be simpler! (I will leave the simplification as an exercise to the reader)

**Review by jameshochadel — CHANGES_REQUESTED** [review]

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java (hunk 4 above)
Comment (jameshochadel) [reply src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Update: See http://www.kauailabs.com/public_files/navx-mxp/apidocs/java/ -> AHRS -> getYaw; return value is in range (-180.0, 180.0).

File: src/org/usfirst/frc/team2399/robot/OI.java [hunk 14]
```diff
@@ -57,6 +65,10 @@ public static double getRightY() {
 	private static Button climbUpButt = new JoystickButton(leftJoy, 86);
 	private static Button climbDownButt = new JoystickButton(leftJoy, 59);
 	private static Button climbStopButt = new JoystickButton(leftJoy, 300);
+	private static Button driveDistanceIncrementButt = new JoystickButton(shooterJoy, 5);
+	private static Button driveDistanceDecrementButt = new JoystickButton(shooterJoy, 3);
+	private static Button driveAngleIncrementButt = new JoystickButton(shooterJoy, 6);
+	private static Button driveAngleDecrementButt = new JoystickButton(shooterJoy, 4);
```
Comment (wmarshall) [inline src/org/usfirst/frc/team2399/robot/OI.java]: Why are we adjusting drive constants on the shooter joystick? Not a fail, just a question.

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java (hunk 5 above)
Comment (wmarshall) [reply src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Ah. Shortest-path angle measurement. There's a way to do this with one conditional, but it's very unintuitive unless you've written this method for a bunch of different projects. Would be happy to help with this.

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java (hunk 6 above)
Comment (wmarshall) [reply src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: This is actually okay. The reason that this is done here vs on the talons is that the control loop specified by this method deals with steering (left/right) while the talons are only aware of velocity (forward/back). Steering is an abstraction over velocity, so it makes sense to sit higher in the abstraction pyramid

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java [hunk 15]
```diff
@@ -178,18 +184,108 @@ public double returnRightDistanceError(){
...
+		}
+		
+		return newDesiredAngle - getCurrentAngle();
+	}
+	
+	public void moveToAngle()
+	{
+		double pOutput = calculateAngleError() * anglePConstant;
+		driveRightVelocity(-pOutput);
+		driveLeftVelocity(pOutput);
```
Comment (wmarshall) [inline src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Why the +/-? what coordinate system are we using? Looks like CW = + , but this isn't documented anywhere.

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java (hunk 8 above)
Comment (wmarshall) [reply src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Same answer as above

**Review by wmarshall — CHANGES_REQUESTED** [review]
This still needs some work. Not much, in that I think we should move it towards testing ASAP, to validate that our NavX works, but enough that it should not be merged in current state.

File: src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java (hunk 6 above)
Comment (jameshochadel) [reply src/org/usfirst/frc/team2399/robot/subsystems/DriveTrain.java]: Yep, I checked in with Chloe and that makes sense. (I really wish these comments had a 'resolve' button)
